cpp/src/neighbors/scann/detail/scann_build.cuh

```diff
@@ -0,0 +1,383 @@
+                                     params.soar_lambda);
+
+    // Compute and quantize residuals
+    auto avq_residuals = compute_residuals<T, uint32_t>(
+      res, batch_view, raft::make_const_mdspan(centroids_view), batch_labels_view);
+
+    auto avq_quant = quantize_residuals<T, IdxT, uint32_t>(
+      res, raft::make_const_mdspan(avq_residuals.view()), full_codebook.view(), pq_params);
+
+    // Compute and quantize SOAR residuals
+    auto soar_residuals =
+      compute_residuals<T, uint32_t>(res,
+                                     batch_view,
+                                     raft::make_const_mdspan(centroids_view),
+                                     raft::make_const_mdspan(batch_soar_labels_view));
+
+    auto soar_quant = quantize_residuals<T, IdxT, uint32_t>(
+      res, raft::make_const_mdspan(soar_residuals.view()), full_codebook.view(), pq_params);
+
+    // unpack codes
+    auto quantized_residuals =
+      raft::make_device_matrix<uint8_t, IdxT>(res, batch.size(), num_subspaces);
+    auto quantized_soar_residuals =
+      raft::make_device_matrix<uint8_t, IdxT>(res, batch.size(), num_subspaces);
+
+    unpack_codes(res,
+                 quantized_residuals.view(),
+                 raft::make_const_mdspan(avq_quant.view()),
+                 params.pq_bits,
+                 num_subspaces);
+    unpack_codes(res,
+                 quantized_soar_residuals.view(),
+                 raft::make_const_mdspan(soar_quant.view()),
+                 params.pq_bits,
+                 num_subspaces);
+
+    // quantize dataset to bfloat16, if enabled. Similar to SOAR, quantization
+    // is performed in this loop to improve locality
+    // TODO (rmaschal): Might be more efficient to do on CPU, to avoid DtoH copy
+
+    auto bf16_dataset = raft::make_device_matrix<int16_t, int64_t>(res, batch_view.extent(0), dim);
+
+    if (params.bf16_enabled) {
+      raft::linalg::map_offset(res, bf16_dataset.view(), [batch_view, dim] __device__(size_t i) {
+        int64_t row_idx = i / dim;
+        int64_t col_idx = i % dim;
+
+        nv_bfloat16 val = __float2bfloat16(batch_view(row_idx, col_idx));
+
+        return reinterpret_cast<int16_t&>(val);
+      });
+    }
+
+    // Prefetch next batch
+    dataset_vec_batches.prefetch_next_batch();
+
+    // Copy unpacked codes to host
+    // TODO (rmaschal): these copies are blocking and not overlapped
+    auto h_quantized_residuals_view = raft::make_host_matrix_view<uint8_t, uint32_t>(
+      idx.quantized_residuals().data_handle() + batch.offset() * num_subspaces,
+      batch.size(),
+      num_subspaces);
+
+    raft::copy(h_quantized_residuals_view.data_handle(),
+               quantized_residuals.data_handle(),
+               quantized_residuals.size(),
+               stream);
+
+    auto h_quantized_soar_residuals_view = raft::make_host_matrix_view<uint8_t, uint32_t>(
+      idx.quantized_soar_residuals().data_handle() + batch.offset() * num_subspaces,
```

Comment:
We could simplify like this. But do we even need this view? We could just put the same line into `raft::copy`
```suggestion
      &(idx.quantized_soar_residuals())(batch.offset(), 0),
```